manifests/params.pp

```diff
@@ -76,6 +76,10 @@
   # Custom puppetmaster
   $puppetmaster              = $::puppetmaster
 
+  # Hashes containing additional settings
+  $additional_settings       = {}
```

Comment:
My pattern is to abstract the config file structure in Puppet but map kv pairs in hashes. As such "main", "agent" and "master" would be the ideal user UX. I think if you want to go all the way, that would be overall the best approach. But beware, I'm in the crazy Hiera-model camp ;)